Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm

```diff
@@ -262,41 +292,113 @@ static bool isCopyDisplayedPixelBufferAvailable()
-    if (synchronizerTime > durationMediaTime())
-        return durationMediaTime();
+    if (synchronizerTime < m_lastSeekTime)
+        return m_lastSeekTime;
 
     return synchronizerTime;
 }
 
 void MediaPlayerPrivateWebM::seekToTarget(const SeekTarget& target)
 {
-    ALWAYS_LOG(LOGIDENTIFIER, "target = ", target);
-    [m_synchronizer setRate:0 time:PAL::toCMTime(target.time)];
+    ALWAYS_LOG(LOGIDENTIFIER, "time = ", target.time, ", negativeThreshold = ", target.negativeThreshold, ", positiveThreshold = ", target.positiveThreshold);
+
+    m_pendingSeek = target;
+
+    if (m_seekTimer.isActive())
+        m_seekTimer.stop();
+    m_seekTimer.startOneShot(0_s);
+}
+
+void MediaPlayerPrivateWebM::seekInternal()
+{
+    if (!m_pendingSeek)
+        return;
+
+    auto pendingSeek = std::exchange(m_pendingSeek, { }).value();
+    m_lastSeekTime = pendingSeek.time;
+    [m_synchronizer setRate:0 time:PAL::toCMTime(m_lastSeekTime)];
+
+    m_seekState = Seeking;
+
+    MediaTime synchronizerTime = PAL::toMediaTime([m_synchronizer currentTime]);
+
+    m_synchronizerSeeking = synchronizerTime != m_lastSeekTime;
+    ALWAYS_LOG(LOGIDENTIFIER, "seekedTime = ", m_lastSeekTime, ", synchronizerTime = ", synchronizerTime, "synchronizer seeking = ", m_synchronizerSeeking);
+
+    if (!m_synchronizerSeeking) {
+        // In cases where the destination seek time precisely matches the synchronizer's existing time
+        // no time jumped notification will be issued. In this case, just notify the MediaPlayer that
+        // the seek completed successfully.
+        maybeCompleteSeek();
+        return;
+    }
+
+    flush();
+
     for (auto& trackBufferPair : m_trackBufferMap) {
         TrackBuffer& trackBuffer = trackBufferPair.second;
         auto trackId = trackBufferPair.first;
 
         trackBuffer.setNeedsReenqueueing(true);
-        reenqueueMediaForTime(trackBuffer, trackId, target.time);
+        reenqueueMediaForTime(trackBuffer, trackId, m_lastSeekTime);
     }
-    [m_synchronizer setRate:m_rate];
+
+    maybeCompleteSeek();
+}
+
+void MediaPlayerPrivateWebM::maybeCompleteSeek()
+{
+    if (m_seekState == SeekCompleted)
+        return;
+    if (hasVideo() && !m_hasAvailableVideoFrame) {
+        ALWAYS_LOG(LOGIDENTIFIER, "waiting for video frame");
+        m_seekState = WaitingForAvailableFame;
+        return;
+    }
+    m_seekState = Seeking;
+    ALWAYS_LOG(LOGIDENTIFIER);
+    if (m_synchronizerSeeking) {
+        ALWAYS_LOG(LOGIDENTIFIER, "Synchronizer still seeking, bailing out");
+        return;
+    }
+    m_seekState = SeekCompleted;
+    if (shouldBePlaying())
+        [m_synchronizer setRate:m_rate];
     if (auto player = m_player.get()) {
-        player->seeked(target.time);
+        player->seeked(m_lastSeekTime);
         player->timeChanged();
     }
 }
 
+bool MediaPlayerPrivateWebM::seeking() const
+{
+    return m_pendingSeek || m_synchronizerSeeking || m_seekState != SeekCompleted;
```

Comment:
It is not totally clear why we need m_synchronizerSeeking given we have m_seekState.
Can we remove m_synchronizerSeeking?

Reply:
It would mean to redo/review all the existing logic , I don't want to do it at this stage.

Reply:
ok, I misread your comment, I thought you were referring to the entire code, not just this method. yes will do